CODE_STYLE.md

```diff
@@ -0,0 +1,3293 @@
+  It's more common and obviously more concise.
+<sup>[[link](#ternary-operator)]</sup>
+
+  ```Ruby
+  # bad
+  result = if some_condition then something else something_else end
+
+  # good
+  result = some_condition ? something : something_else
+  ```
+
+* <a name="no-semicolon-ifs"></a>
+  Do not use `if x; ...`. Use the ternary operator instead.
+<sup>[[link](#no-semicolon-ifs)]</sup>
+
+  ```Ruby
+  # bad
+  result = if some_condition; something else something_else end
+
+  # good
+  result = some_condition ? something : something_else
+  ```
+
+* <a name="bang-not-not"></a>
+  Use `!` instead of `not`.
+<sup>[[link](#bang-not-not)]</sup>
+
+  ```Ruby
+  # bad - braces are required because of op precedence
+  x = (not something)
+
+  # good
+  x = !something
+  ```
+
+* <a name="no-bang-bang"></a>
+  Avoid the use of `!!`.
+<sup>[[link](#no-bang-bang)]</sup>
+
+  ```Ruby
+  # bad
+  x = 'test'
+  # obscure nil check
+  if !!x
+    # body omitted
+  end
+
+  x = false
+  # double negation is useless on booleans
+  !!x # => false
+
+  # good
+  x = 'test'
+  unless x.nil?
+    # body omitted
+  end
+  ```
+
+* <a name="no-and-or-or"></a>
+  No usage of the `and` and `or` keywords is accepted. Always use `&&` and `||` instead.
+<sup>[[link](#no-and-or-or)]</sup>
+
+  ```Ruby
+  # bad
+  # boolean expression
+  if some_condition and some_other_condition
+    do_something
+  end
+
+  # control flow
+  document.saved? or document.save!
+
+  # good
+  # boolean expression
+  if some_condition && some_other_condition
+    do_something
+  end
+
+  # control flow
+  document.saved? || document.save!
+  ```
+
+* <a name="no-multiline-ternary"></a>
+  Avoid multi-line `?:` (the ternary operator); use `if/unless` instead.
+<sup>[[link](#no-multiline-ternary)]</sup>
+
+* <a name="if-as-a-modifier"></a>
+  Favor modifier `if/unless` usage when you have a single-line body. Another
+  good alternative is the usage of control flow `&&/||`.
+<sup>[[link](#if-as-a-modifier)]</sup>
+
+  ```Ruby
+  # bad
+  if some_condition
+    do_something
+  end
+
+  # good
+  do_something if some_condition
+
+  # another good option
+  some_condition && do_something
+  ```
+
+* <a name="no-multiline-if-modifiers"></a>
+  Avoid modifier `if/unless` usage at the end of a non-trivial multi-line
+  block.
+<sup>[[link](#no-multiline-if-modifiers)]</sup>
+
+  ```Ruby
+  # bad
+  10.times do
+    # multi-line body omitted
+  end if some_condition
+
+  # good
+  if some_condition
+    10.times do
+      # multi-line body omitted
+    end
+  end
+  ```
+
+* <a name="unless-for-negatives"></a>
+
+Avoid usage of `unless` over negative conditions (or control flow), it
+can cause issues when having multi term expressions.
+<sup>[[link](#unless-for-negatives)]</sup>
+
+  ```Ruby
+  # bad
+  do_something unless some_condition
+
+  # good
+  do_something if !some_condition
+
+  # another good option
+  some_condition || do_something
```

Comment:
See https://github.com/jordansissel/experiments/tree/master/ruby/logger-string-vs-block for my prior research on this. 

Running today on my laptop:

```
% ruby rubylogger.rb
         string | jruby/  1.9.3 |     0.50
      string_if | jruby/  1.9.3 |     0.13
     string_and | jruby/  1.9.3 |     0.13
     control_if | jruby/  1.9.3 |     0.12
    control_and | jruby/  1.9.3 |     0.12
          block | jruby/  1.9.3 |     0.26
    doubleblock | jruby/  1.9.3 |     0.28
```

I don't think it tests `logger.info("string") { some_hash_context_here }` but it does test blocks.

Specifically, it tries to test the case where something is _not_ needing to be logged and we want to avoid extra work.